### PR TITLE
2291249 Modal display

### PR DESCRIPTION
**Pull request by kmoll**

### CODE REVIEW

File: entity_browser.services.yml [hunk 1]
```diff
@@ -13,4 +13,6 @@ services:
     parent: default_plugin_manager
   entity_browser.route_subscriber:
     class: Drupal\entity_browser\RouteSubscriber
+    tags: 
+     - { name: event_subscriber }
```
Comment (slashrsm) [inline entity_browser.services.yml]: Sure we need that? I checked Drupal\Core\DependencyInjection\Compiler\RegisterKernelListenersPass and it seems to rely on getSubscribedEvents() which we don't implement.

File: js/entity_browser.modal.js [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        }
+
+        if (typeof fn === 'function') {
+          $('a[data-uuid="' + drupalSettings.entity_browser.modal.uuid + '"]').not('.entity-browser-processed')
+          .bind('entities-selected', fn).addClass('entity-browser-processed');
+        }
+      }
+    }
+  }
+}(jQuery, Drupal, drupalSettings));
```
Comment (slashrsm) [inline js/entity_browser.modal.js]: Missing newline.

File: src/Controllers/FormAjaxController.php [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+        ]);
+  
+    // The form needs to be processed; prepare for that by setting a few
+    // internal variables.
+    $form_state->setUserInput($request->request->all());
+    $form_id = $form['#form_id'];
+  
+    return new FileAjaxForm($form, $form_state, $form_id, $form_build_id, $commands);
+  }
+}
```
Comment (slashrsm) [inline src/Controllers/FormAjaxController.php]: Missing newline.

File: entity_browser.services.yml (hunk 1 above)
Comment (kmoll) [inline entity_browser.services.yml]: I added this in order for our RouteSubscriber to extend the RouteSubscriberBase class that allows us to alter routes.  I had to do this in order to override the FormAjaxController::content method to call our Class.

File: src/Controllers/FormAjaxController.php [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+use Symfony\Component\HttpFoundation\Request;
+use Drupal\Core\Ajax\UpdateBuildIdCommand;
+use Drupal\Core\Form\FormState;
+use Drupal\Core\DependencyInjection\ContainerInjectionInterface;
+use Drupal\system\FileAjaxForm;
+use Drupal\Core\Form\FormBuilderInterface;
+use Drupal\Core\Entity\EntityManagerInterface;
+
+class FormAjaxController extends \Drupal\system\Controller\FormAjaxController {
```
Comment (slashrsm) [inline src/Controllers/FormAjaxController.php]: You can do 

```
use \Drupal\system\Controller\FormAjaxController as CoreFormAjaxController;
```

so you don't need to use absolute form with extend.

File: src/Controllers/FormAjaxController.php [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+    $form_state = new FormState();
+    $form_build_id = $request->request->get('form_build_id');
+  
+    $form_id = $request->request->get('form_id');
+    
+    if (!substr($form_id, 0, 15) == 'entity_browser') {
+      return parent::getForm($request);
+    }
+    
+    // Get the form from the cache.
```
Comment (slashrsm) [inline src/Controllers/FormAjaxController.php]: We ~~are~~ should be using this only with entity browser which means all code below this point is always unused. Let's remove it.

File: src/RouteSubscriber.php [hunk 6]
```diff
@@ -90,5 +91,13 @@ protected function getBrowserIDsWithRoute() {
       ->condition("display", $ids, 'IN')
       ->execute();
   }
+  
+  public function alterRoutes(RouteCollection $collection) {
+    if ($route = $collection->get('system.ajax')) {
```
Comment (slashrsm) [inline src/RouteSubscriber.php]: It is not a good idea to disable cache for the entire site. It would be better to create new route and tell form to use it. Not sure if it is possible to do that, but I doubt it is hardcoded. In case it is we need to submit a patch against core.

File: src/Entity/EntityBrowser.php [hunk 7]
```diff
@@ -393,11 +395,15 @@ public function buildForm(array $form, FormStateInterface $form_state) {
       '#type' => 'value',
       '#value' => array_map(function(EntityInterface $item) {return $item->id();}, $this->getSelectedEntities())
     );
-
+
```
Comment (slashrsm) [inline src/Entity/EntityBrowser.php:401]: Whitespace.

File: src/Entity/EntityBrowser.php [hunk 8]
```diff
@@ -381,6 +382,7 @@ public function getFormId() {
    * {@inheritdoc}
    */
   public function buildForm(array $form, FormStateInterface $form_state) {
+
```
Comment (slashrsm) [inline src/Entity/EntityBrowser.php:396]: Newline probably uneeded.

File: src/Plugin/EntityBrowser/Display/Modal.php [hunk 9]
```diff
@@ -0,0 +1,238 @@
...
+  /**
+   * {@inheritdoc}
+   */
+  public function addAjax(array &$form) {
+    // Add the browser id to use in the FormAjaxController.
+    $form['browser_id'] = array(
+      '#type' => 'hidden',
+      '#value' => $this->configuration['entity_browser_id'],
+    );
+    $form[$form['#browser_parts']['widget']]['actions']['submit']['#ajax'] = array(
```
Comment (slashrsm) [inline src/Plugin/EntityBrowser/Display/Modal.php]: This assumed how widget's form looks like. Is there another, more generic, way to do this?

File: src/DisplayAjaxInterface.php [hunk 10]
```diff
@@ -0,0 +1,26 @@
+<?php
+/**
+ * @file
+ * Contains \Drupal\entity_browser\DisplayAjaxInterface.
```
Comment (slashrsm) [inline src/DisplayAjaxInterface.php:4]: DisplayRouterInterface lives in Plugin\EntityBrowser\Display. It probably makes sense for them to live in the same place. Let's move this class there or DisplayRouterInterface here.

File: js/entity_browser.modal.js [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+(function ($, Drupal, drupalSettings) {
+	
+  "use strict";
+  
+  Drupal.AjaxCommands.prototype.select_entities = function (ajax, response, status) {
+    var uuid = drupalSettings.entity_browser.modal.uuid;
+    $('a[data-uuid="' + uuid + '"]').trigger('entities-selected', [uuid, response.entities])
+    .removeClass('entity-browser-processed').unbind('entities-selected');
+  };
+
```
Comment (slashrsm) [inline js/entity_browser.modal.js:16]: 2 newlines.

File: src/Plugin/EntityBrowser/Display/Modal.php (hunk 9 above)
Comment (kmoll) [inline src/Plugin/EntityBrowser/Display/Modal.php]: yeah, I was thinking about that as well.  I guess I will have to loop through the form with element_children, and find the submit element?  Right now thats the only generic way I can think if, but if there are more than one submit it could be tricky.  Any better ideas?  Can implement some sort of convention that needs to be followed in order for it to work?

File: src/RouteSubscriber.php (hunk 6 above)
Comment (kmoll) [inline src/RouteSubscriber.php]: Ok, I originally had code in our Controller that just passed back the original if it wasn't our form, but you are right.  I think its better to see if we can create a new route and use the 'path' element on the ajax to call our own route instead of system.ajax.

File: src/DisplayAjaxInterface.php (hunk 10 above)
Comment (kmoll) [inline src/DisplayAjaxInterface.php:4]: I'm thinking that we move the DisplayRouterInterface.php to the src root.  The rest of the display parent classes and interfaces are there, thoughts?

Comment (slashrsm) [inline src/DisplayAjaxInterface.php:4]: OK.

File: entity_browser.routing.yml [hunk 12]
```diff
@@ -1,2 +1,11 @@
 route_callbacks:
   - 'entity_browser.route_subscriber:routes'
+
```
Comment (slashrsm) [inline entity_browser.routing.yml]: whitespace

File: src/Plugin/EntityBrowser/Display/Modal.php [hunk 13]
```diff
@@ -0,0 +1,241 @@
...
+          'data-dialog-options' => json_encode(
+            array(
+              'width' => $this->configuration['width'],
+              'height' => $this->configuration['height'],
+            )
+          ),
+          'data-accepts' => "application/vnd.drupal-modal",
+        ],
+        '#attached' => [
+          'js' => [
```
Comment (slashrsm) [inline src/Plugin/EntityBrowser/Display/Modal.php]: This throws 

> LogicException: You are not allowed to use js in #attached in drupal_process_attached()

Needs to be chnaged to drupalSettings. See https://www.drupal.org/node/2383115

File: src/Plugin/EntityBrowser/Display/Modal.php [hunk 14]
```diff
@@ -0,0 +1,241 @@
...
+   */
+  public function addAjax(array &$form) {
+    // Add the browser id to use in the FormAjaxController.
+    $form['browser_id'] = array(
+      '#type' => 'hidden',
+      '#value' => $this->configuration['entity_browser_id'],
+    );
+    // @TODO find more generic way of doing this.
+    $form['actions']['submit']['#ajax'] = array(
+      'callback' => array($this, 'widgetAjaxCallback'),
```
Comment (slashrsm) [inline src/Plugin/EntityBrowser/Display/Modal.php:172]: $this is added to form structure here. Form is serialized and cached on AJAX requests, which ends up in fatals

> User error: The container was serialized. in Drupal\Core\DependencyInjection\Container->__sleep() (line 34 of /var/www/drupal8/core/lib/Drupal/Core/DependencyInjection/Container.php).

and 

> BadMethodCallException: Settings cannot be serialized. in Drupal\Core\Site\Settings->__sleep() (line 66 of /var/www/drupal8/core/lib/Drupal/Core/Site/Settings.php).

In order to prevent this you need to implement __sleep() and __wakeup() magic methods on this class. See EntityBrowser.php for an example. Also see https://github.com/drupal-media/entity_browser/pull/39 which re-enables form cache and tries to fix problems that I noticed related to that. It would be great if we could test this code with that PR merged in.

File: src/Plugin/EntityBrowser/Widget/View.php [hunk 15]
```diff
@@ -62,13 +62,6 @@ public function getForm(array &$original_form, FormStateInterface $form_state) {
       }
     }
 
-    $form['actions'] = [
```
Comment (slashrsm) [inline src/Plugin/EntityBrowser/Widget/View.php:65]: We need to remove from Upload.php too. Label is different there. We should probably make label configurable and/or allow widgets to override it.
